core/src/main/scala/org/bitcoins/core/crypto/TxSigComponent.scala

```diff
@@ -166,6 +180,17 @@ object BaseTxSigComponent {
 
 }
 
+object P2SHTxSigComponent {
+
+  def apply(
+      transaction: Transaction,
+      inputIndex: UInt32,
+      output: TransactionOutput,
+      flags: Seq[ScriptFlag]): BaseTxSigComponent = {
+    BaseTxSigComponent(transaction, inputIndex, output, flags)
```

Comment:
You should probably define a `private case class P2SHTxSigComponentImpl(...)` and use it as a constructor here.

Reply:
I'm actually hoping to slowly get rid of that pattern as it obfuscates the invariants imo as its only purpose is to restrict construction where I think we should just make things correct by construction

Reply:
That said something has to be concrete and i see no reason not to make `P2SHTxSigComponent` concrete

Reply:
I think it makes more sense to do Chris's suggestion here because that is what is done throughout `TxSigCompenent.scala` and in a later refactor we can change it